### PR TITLE
chore: remove duplicated retry logic from quickwit-aws

### PR DESCRIPTION
**Pull request by etolbakov**
### Description

The scope of a change aims to remove duplicated retry logic from `quickwit-aws` and reuse it from `quickwit-common`

### How was this PR tested?
Tests in `quickwit-aws` are passing after the change has been made

Related issues:
https://github.com/quickwit-oss/quickwit/issues/3756

### CODE REVIEW

**Comment by etolbakov** [comment]
Hello @guilload 
I was curious to try this issue. Could you please take a look at the PR when you have time?
---
Regards, Eugene

**Comment by guilload** [comment]
Hi Eugene, I'd like to take the refactor a step further by moving the `retry_with_mockable_time` function and the associated test to `quickwit-common::retry`.

**Comment by etolbakov** [comment]
Sure, let me give it a go.

**Comment by etolbakov** [comment]
Hello Adrien,
I've addressed the requirement to move the `retry_with_mockable_time` and the associated test to `quickwit-common::retry`. 
However, I came across the issue which I can not solve without a bit of guidance. 
I appreaciate that for an experienced developer an explaination may take more time than the actual implementation, so please let me know if that the case.

The issue:
1) using `Retryable` from `common` produces the following problem in [error.rs](https://github.com/etolbakov/quickwit/blob/chore/quickwit-aws-retry/quickwit/quickwit-aws/src/error.rs)

```
only traits defined in the current crate can be implemented for types defined outside of the crate
```
2) that was the reason why `RetryableWrapper` was introduced.
3) but that fires back in `s3_compatible_storage.rs`, f.e.

```
error[E0277]: the trait bound `HeadObjectError: Retryable` is not satisfied
   --> quickwit-storage/src/object_storage/s3_compatible_storage.rs:819:15
    |
819 |         .await?;
    |               ^ the trait `Retryable` is not implemented for `HeadObjectError`
    |
    = help: the following other types implement trait `Retryable`:
              RetryableWrapper<SdkError<E>>
              RetryableWrapper<AbortMultipartUploadError>
              RetryableWrapper<CompleteMultipartUploadError>
              RetryableWrapper<CreateMultipartUploadError>
              RetryableWrapper<DeleteObjectError>
              RetryableWrapper<DeleteObjectsError>
              RetryableWrapper<GetObjectError>
              RetryableWrapper<HeadObjectError>
            and 3 others
```
which totally make sense... and this is where I get stuck. 
Could you please give me a steer?

**Comment by guilload** [comment]
It's tricky, that's why I did not perform this refactor last time :)

You have to define a new `AwsRetryable` trait in the `aws` crate and wrapper then define do something like that:
```rust
pub trait AwsRetryable {
    fn is_retryable(&self) -> bool {
        false
    }
}

#[derive(Debug)]
struct AwsRetryableWrapper<E>(E);

impl<E> Retryable for AwsRetryableWrapper<E> where E: AwsRetryable {
    fn is_retryable(&self) -> bool {
        self.0.is_retryable()
    }
}

pub async fn aws_retry<U, E, Fut>(retry_params: &RetryParams, f: impl Fn() -> Fut) -> Result<U, E>
where
    Fut: Future<Output = Result<U, E>>,
    E: AwsRetryable + Debug + 'static,
{
    retry(retry_params, || f().map_err(AwsRetryableWrapper)).await.map_err(|error| error.0)
}
```

I suspect it won't be enough because the error handling in `storage` is quite complex and could use a refactor as well :/ For instance, I suspect `Retryable` bound in this `From` implementation is unnecessary and might get in our way:
```rust
impl<E> From<SdkError<E>> for StorageError
where E: Send + Sync + std::error::Error + 'static + ToStorageErrorKind + Retryable
```

We certainly have issues that are more rewarding to work on.

**Comment by etolbakov** [comment]
Thanks for explaining. Will proceed with experiments.

**Comment by etolbakov** [comment]
Hi Adrien,
Could you please take a look at the this attempt when you have time?

**Comment by guilload** [comment]
You can run `make fmt` to fix the formatting issues.
You can run `make fix` to fix some of the linting issues.
You can run `make fmt` to test the entire project.

You can run `cargo test --manifest-path quickwit/Cargo.toml -p quickwit-common` to test a single crate.

-> [Contributing guidelines](https://github.com/quickwit-oss/quickwit/blob/main/CONTRIBUTING.md)

**Comment by guilload** [comment]
@etolbakov, please drop this commit (b2c2e44924de6b52af2264eb03f614531b290c52), on top of your branch it should bring this PR where we want it to be.

**Comment by etolbakov** [comment]
Hi Adrien,
thank you very much for the help!

Re: formatting & linting issues - noted, sorry still haven't developed a habit.

Re: https://github.com/quickwit-oss/quickwit/commit/45744bf9850ba68038e3c2469ad45a1ef0034784 - I've moved that to my branch. Need to understand how far I was from the final solution.

**Comment by guilload** [comment]
You were there. Functionally, reusing `compute_delay` was the only thing missing. The rest was just me doing some additional cleanup and small refactors.

File: quickwit/quickwit-storage/src/object_storage/s3_compatible_storage.rs [hunk 1]
```diff
@@ -54,7 +55,7 @@ use crate::{
 };
 
 /// Semaphore to limit the number of concurent requests to the object store. Some object stores
-/// (R2, SeaweedFs...) return errors when too many concurent requests are emitted.
+/// (R2, SeaweedFs...) return errors when too many concurrent requests are emitted.
```
Comment (guilload) [inline quickwit/quickwit-storage/src/object_storage/s3_compatible_storage.rs:58]: Thanks for fixing the typos.

**Review by guilload — APPROVED** [review]

**Comment by guilload** [comment]
Merged. Thanks @etolbakov !

**Comment by etolbakov** [comment]
Thanks Adrien for your help and guidance! 
That was a good learning exercise!
> We certainly have issues that are more rewarding to work on.
please let me know if there's anything particular on your mind so I can take a look

File: quickwit/quickwit-storage/src/object_storage/s3_compatible_storage.rs (hunk 1 above)
Comment (etolbakov) [reply quickwit/quickwit-storage/src/object_storage/s3_compatible_storage.rs:58]: try leaving the place in slightly better shape before found it 😃

**Pull request merged**